### PR TITLE
Rate generators session pool workstealing

### LINKED ISSUE #359: Improve Session pool
The comments at https://github.com/Hyperfoil/Hyperfoil/blob/hyperfoil-all-0.26/core/src/main/java/io/hyperfoil/core/impl/ElasticPoolImpl.java#L27-L28 and https://github.com/Hyperfoil/Hyperfoil/pull/357#issuecomment-2156186667 shows that we have different ways to improve the `ElasticPool` powering the Session's pool.

The easiest solution has been attempted at https://github.com/franz1981/Hyperfoil/commit/25948a82e02ec30e09cd8f5e3b96f14ee4279c50 which should already improve the current situation, but we need a jmh module to prove it.

### PR DESCRIPTION
**Pull request by franz1981**
Fixes #359 

## Changes proposed

- [x] Implemented basic striped pool implementation
- [X] Adding unit tests
- [X] Adding JMH benchmark module
- [x] Leverage Thread-affinity of `Session`s to populate and assign it to dedicated queues

The last point is crucial to better work with the phase instance which acquire `Session`s: instead of relying on a random thread affinity (assigned in round robin) - we should just use the initial provided one (thinking about the `Session`s specific use case here) in order to both reduce contention and guarantee the thread-local acquired `Session` is executed in the same `Thread` who requested it.
The implication of this choice are not obvious because it means that in a multithreading scenario where the phase instance acquire the Session, it will start it delayed after it complete the `proceed` call, which is something to be evaluated if is the right choice.

### CODE REVIEW

**Comment by franz1981** [comment]
I am not really happy about 4d740c3efe2e7683335078c7997b5ce350a6243c (Which is hanging the CI) mostly because of a quick analysis:
1. phase instances start seems to assume it runs immediately 
2. Such method is not always called from event loops (on simulation start)
3. It breaks an event loop confined pool like the one in this pr 
4. Forcing it to start on the event loop just breaks the CI with test failures - which get several NPE here and there

I have to dig into the root cause and if the assumption on start is there for some reason, or just relax the new pool requirements - but is a shame because it is for a single acquire request.

**Comment by franz1981** [comment]
I've still to improve the `reserve` performance and allocation rate (by far, actually!)
I've added a microbenchmark just for it, which I still have to analyze and improve, but code-wise this is finally ready for review

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        var localQueues = this.localQueues;
+        var session = localQueues[agentThreadId].poll();
+        if (session != null) {
+            incrementUsed();
+            return session;
+        }
+        if (localQueues.length == 1) {
+            return null;
+        }
+        return acquireFromOtherLocalQueues(localQueues, agentThreadId);
```
Comment (lampajr) [inline core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:107]: Couple of questions here:
- Why do we need explicit unboxing for `boxedAgentThreadId`? isn't it already automatically managed?
- Why do we need to save `this.localQueues` to a local variable? is there any particular reason or benefit?

If there is no particular reason fo rboth doubts I think we can simplify a little bit:
```suggestion
        var boxedAgentThreadId = localAgentThreadId.get();
        if (boxedAgentThreadId == null) {
            return acquireFromLocalQueues();
        }
        var session = localQueues[boxedAgentThreadId].poll();
        if (session != null) {
            incrementUsed();
            return session;
        }
        if (this.localQueues.length == 1) {
            return null;
        }
        return acquireFromOtherLocalQueues(this.localQueues, boxedAgentThreadId);
```

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+    }
+
+    private Session acquireFromLocalQueues() {
+        int workStealIndex = nextWorkStealIndex();
+        int currentIndex = workStealIndex;
```
Comment (lampajr) [inline core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:112]:
```suggestion
        int currentIndex = nextWorkStealIndex();
```

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            if (newSession == null) {
+                throw new IllegalStateException("Session supplier returned null");
+            }
+            var executor = newSession.executor();
+            var boxedAgentThreadId = agentThreadIdPerExecutor.get(executor);
+            if (boxedAgentThreadId == null) {
+                throw new IllegalStateException("This session is associated with an unknown EventExecutor");
+            }
+            // sanity check against the sessions agentThreadId
+            if (boxedAgentThreadId.intValue() != newSession.agentThreadId()) {
```
Comment (lampajr) [inline core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:229]:
```suggestion
            if (boxedAgentThreadId != newSession.agentThreadId()) {
```

**Review by lampajr — COMMENTED** [review]
Hey @franz1981 this looks really great!

I did a first round and I found just some minors/simplification. I'll let you decide whether to include them or not.

I see you explicitly box/unbox integers, is there any particular reason for doing that?

File: core/src/main/java/io/hyperfoil/core/impl/PhaseInstanceImpl.java [hunk 4]
```diff
@@ -264,13 +265,23 @@ protected boolean startNewSession() {
...
       return true;
    }
 
+   private void noSessionsAvailable() {
+      if (failedSessionAcquisitionAction != null) {
+          failedSessionAcquisitionAction.run();
+      }
+      if (!sessionLimitExceeded) {
+         sessionLimitExceeded = true;
+      }
```
[the remark this answers is not in the document]
Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/PhaseInstanceImpl.java:281]: The problem is that it is a costy operation to volatile set (it waste like hundreds of cycles more than a "plain" non volatile set") - so I suppose that Radim decided to pay this just the first time (unless reset happen)

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java (hunk 1 above)
Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:107]: The main reason was to avoid twice `Integer::intValue`: given that there are atomic operations in-between the 2 unboxing operations, the JIT can decide to actual perform it twice, without trusting that Integer's value hasn't changed :/

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java (hunk 3 above)
Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:229]: I don't remember by heart if the left side in unboxed or the right side is boxed, so I preferred to be safe and avoid a potential allocation path.

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                throw new IllegalStateException("Session::getAgentThreadId is not associated with the provided EventExecutor");
+            }
+            var agentThreadId = boxedAgentThreadId.intValue();
+            if (newSessionsPerEventExecutors[agentThreadId] == null) {
+                newSessionsPerEventExecutors[agentThreadId] = new ArrayDeque<>(perEventExecutorCapacity);
+            }
+            newSessionsPerEventExecutors[agentThreadId].add(newSession);
```
[the remark this answers is not in the document]
Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:236]: reserving usually happen at the very begining of the application: i've saved unboxing to happen few times while the code is still cold, to save the potential `invokespecial` bytecode to unbox, to happen 3 more times, per session created

File: core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java (hunk 2 above)
Comment (franz1981) [reply core/src/main/java/io/hyperfoil/core/impl/AffinityAwareSessionPool.java:112]: thanks, it was a leftover of the previous implamentation, good catch!

**Comment by franz1981** [comment]
@lampajr I've yet to improve the reserve benchmark because I'm paying the class loading cost there :/ 
I have to try wrk2 (which uses a LOT of sessions, for high rates) and see if the diff in GC time is measurable at startup...
Right now given that we uses different queues (which are way bigger(s)) than `ArrayBlockingQueue` it is expected, but don't want to push this that far - just need to make sure that is not too much impacted.

**Comment by franz1981** [comment]
```
$ java -jar target/benchmarks.jar PoolBenchmark -pworkStealingPool=false,true -pcapacity=1024 -t 1 -pfakeEventExecutors=0

Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt   Score   Error   Units
PoolBenchmark.acquireAndRelease                 1        1024                     0       0               false  thrpt   10  11.673 ± 0.013  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0               false  thrpt   10  11.673 ± 0.013  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0               false  thrpt   10  11.673 ± 0.013  ops/us
PoolBenchmark.acquireAndRelease                 1        1024                     0       0                true  thrpt   10  17.461 ± 0.065  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0                true  thrpt   10  17.461 ± 0.065  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0                true  thrpt   10  17.461 ± 0.065  ops/us
```
```
$ java -jar target/benchmarks.jar PoolBenchmark -pworkStealingPool=false,true -pcapacity=1024 -t 2 -pfakeEventExecutors=0

Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt  Score   Error   Units
PoolBenchmark.acquireAndRelease                 1        1024                     0       0               false  thrpt   10  2.628 ± 0.082  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0               false  thrpt   10  2.629 ± 0.082  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0               false  thrpt   10  2.629 ± 0.082  ops/us
PoolBenchmark.acquireAndRelease                 1        1024                     0       0                true  thrpt   10  9.518 ± 0.279  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0                true  thrpt   10  9.518 ± 0.279  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0                true  thrpt   10  9.518 ± 0.279  ops/us
```
```
$ java -jar target/benchmarks.jar PoolBenchmark -pworkStealingPool=false,true -pcapacity=1024 -t 4 -pfakeEventExecutors=0

Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt  Score   Error   Units
PoolBenchmark.acquireAndRelease                 1        1024                     0       0               false  thrpt   10  6.461 ± 0.253  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0               false  thrpt   10  6.461 ± 0.253  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0               false  thrpt   10  6.461 ± 0.253  ops/us
PoolBenchmark.acquireAndRelease                 1        1024                     0       0                true  thrpt   10  8.513 ± 0.190  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0                true  thrpt   10  8.514 ± 0.190  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0                true  thrpt   10  8.514 ± 0.190  ops/us
```
The next benchmark run is obtained by configuring 8 event executors (4 real + 4 fake) which fairly distribute the configured capacity i.e. 128 / 8 = 16 sessions per event executor
The real threads which acquire/release are just 4, acquiring/releasing 32 sessions per benchmark invocation, which means that 16 of these sessions will belong to the local session while the remaining (16) are acquired from the other queues based on the workstealing policy:
```
$ java -jar target/benchmarks.jar PoolBenchmark -pworkStealingPool=false,true -pcapacity=128 -pburst=32 -t 4 -pfakeEventExecutors=4

Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt  Score   Error   Units
PoolBenchmark.acquireAndRelease                32         128                     4       0               false  thrpt   10  0.205 ± 0.002  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     4       0               false  thrpt   10  6.575 ± 0.063  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     4       0               false  thrpt   10  6.575 ± 0.063  ops/us
PoolBenchmark.acquireAndRelease                32         128                     4       0                true  thrpt   10  0.205 ± 0.005  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     4       0                true  thrpt   10  6.577 ± 0.150  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     4       0                true  thrpt   10  6.577 ± 0.150  ops/us
```
Removing the fake event executors, instead (which means no workstealing is happening):
```
Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt  Score   Error   Units
PoolBenchmark.acquireAndRelease                32         128                     0       0               false  thrpt   10  0.202 ± 0.004  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     0       0               false  thrpt   10  6.449 ± 0.135  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     0       0               false  thrpt   10  6.449 ± 0.135  ops/us
PoolBenchmark.acquireAndRelease                32         128                     0       0                true  thrpt   10  0.280 ± 0.059  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     0       0                true  thrpt   10  8.969 ± 1.900  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     0       0                true  thrpt   10  8.969 ± 1.900  ops/us
PoolBenchmark.acquireAndRelease                32         128                    16       0               false  thrpt   10  0.210 ± 0.002  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                    16       0               false  thrpt   10  6.713 ± 0.057  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                    16       0               false  thrpt   10  6.713 ± 0.057  ops/us
PoolBenchmark.acquireAndRelease                32         128                    16       0                true  thrpt   10  0.220 ± 0.001  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                    16       0                true  thrpt   10  7.032 ± 0.045  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                    16       0                true  thrpt   10  7.032 ± 0.045  ops/us
``` 
Which shows that the workstealing operation is not the main bottleneck here (the counter update, instead, is :"( ) given that the performance hit is not really huge.

For the workstealing use case instead, performance are pretty much on par of the existing pool, but adding GC profiling shows an interesting picture here:
```
Benchmark                                           (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt   Score    Error   Units
PoolBenchmark.acquireAndRelease                          32         128                     4       0               false  thrpt   10   0.213 ±  0.009  ops/us
PoolBenchmark.acquireAndRelease:acquired                 32         128                     4       0               false  thrpt   10   6.822 ±  0.273  ops/us
PoolBenchmark.acquireAndRelease:gc.alloc.rate            32         128                     4       0               false  thrpt   10   5.113 ±  0.350  MB/sec
PoolBenchmark.acquireAndRelease:gc.alloc.rate.norm       32         128                     4       0               false  thrpt   10  25.202 ±  2.729    B/op
PoolBenchmark.acquireAndRelease:gc.count                 32         128                     4       0               false  thrpt   10     ≈ 0           counts
PoolBenchmark.acquireAndRelease:released                 32         128                     4       0               false  thrpt   10   6.822 ±  0.273  ops/us
PoolBenchmark.acquireAndRelease                          32         128                     4       0                true  thrpt   10   0.216 ±  0.022  ops/us
PoolBenchmark.acquireAndRelease:acquired                 32         128                     4       0                true  thrpt   10   6.922 ±  0.719  ops/us
PoolBenchmark.acquireAndRelease:gc.alloc.rate            32         128                     4       0                true  thrpt   10   0.009 ±  0.001  MB/sec
PoolBenchmark.acquireAndRelease:gc.alloc.rate.norm       32         128                     4       0                true  thrpt   10   0.046 ±  0.005    B/op
PoolBenchmark.acquireAndRelease:gc.count                 32         128                     4       0                true  thrpt   10     ≈ 0           counts
PoolBenchmark.acquireAndRelease:released                 32         128                     4       0                true  thrpt   10   6.922 ±  0.719  ops/us
```
which shows that the new pool doesn't allocate in the hot path, which is a plus. 
The reason is that the reentrant lock(s) in the queue requires allocating a wait list under contention, while the new pool doesn't need such.

**Comment by franz1981** [comment]
And fec94bd7522f2fe2969041ae1ff4d5ff45fd84b7 shows that I didn't protected the counters good enough from false-sharing: adding a more paranoid padding (https://stackoverflow.com/questions/29199779/false-sharing-and-128-byte-alignment-padding for some info why) increase a performance in pretty much all cases (yay).

**Comment by franz1981** [comment]
Revised numbers including the new padding scheme, for 

2 Threads:
```
Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt   Score   Error   Units
PoolBenchmark.acquireAndRelease                 1        1024                     0       0               false  thrpt   10   2.584 ± 0.242  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0               false  thrpt   10   2.584 ± 0.242  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0               false  thrpt   10   2.584 ± 0.242  ops/us
PoolBenchmark.acquireAndRelease                 1        1024                     0       0                true  thrpt   10  13.225 ± 0.549  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0                true  thrpt   10  13.225 ± 0.549  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0                true  thrpt   10  13.225 ± 0.549  ops/us
```
4 Threads:
```
Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt   Score   Error   Units
PoolBenchmark.acquireAndRelease                 1        1024                     0       0               false  thrpt   10   6.427 ± 0.050  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0               false  thrpt   10   6.427 ± 0.050  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0               false  thrpt   10   6.427 ± 0.050  ops/us
PoolBenchmark.acquireAndRelease                 1        1024                     0       0                true  thrpt   10  11.781 ± 0.797  ops/us
PoolBenchmark.acquireAndRelease:acquired        1        1024                     0       0                true  thrpt   10  11.782 ± 0.797  ops/us
PoolBenchmark.acquireAndRelease:released        1        1024                     0       0                true  thrpt   10  11.782 ± 0.797  ops/us
```
and, with workstealing

i.e. 
```
$ java -jar target/benchmarks.jar PoolBenchmark -pworkStealingPool=false,true -pcapacity=128 -pburst=32 -t 4 -pfakeEventExecutors=4

Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt  Score   Error   Units
PoolBenchmark.acquireAndRelease                32         128                     4       0               false  thrpt   10  0.196 ± 0.006  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     4       0               false  thrpt   10  6.271 ± 0.186  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     4       0               false  thrpt   10  6.271 ± 0.186  ops/us
PoolBenchmark.acquireAndRelease                32         128                     4       0                true  thrpt   10  0.243 ± 0.008  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     4       0                true  thrpt   10  7.787 ± 0.259  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     4       0                true  thrpt   10  7.787 ± 0.259  ops/us
```
while removing the fake executors (while keeping the burst in order to have a fair comparison)
```
Benchmark                                 (burst)  (capacity)  (fakeEventExecutors)  (work)  (workStealingPool)   Mode  Cnt   Score   Error   Units
PoolBenchmark.acquireAndRelease                32         128                     0       0               false  thrpt   10   0.210 ± 0.009  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     0       0               false  thrpt   10   6.733 ± 0.286  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     0       0               false  thrpt   10   6.733 ± 0.286  ops/us
PoolBenchmark.acquireAndRelease                32         128                     0       0                true  thrpt   10   0.345 ± 0.029  ops/us
PoolBenchmark.acquireAndRelease:acquired       32         128                     0       0                true  thrpt   10  11.053 ± 0.924  ops/us
PoolBenchmark.acquireAndRelease:released       32         128                     0       0                true  thrpt   10  11.053 ± 0.924  ops/us
```
it shows that workstealing still make the new pool to have an edge vs the old one, but NOT having workstealing get the best results - pretty much always doubling performance compared to the original pool.

Right now the sole concern is about `reserve` cost, and I still need to publish numbers about it.
```

**Comment by franz1981** [comment]
After the changes https://github.com/Hyperfoil/Hyperfoil/pull/374/commits/b3b6b1a99d2cb7d63e9d246ab554097f8e63cb24
we can now observe what's going on for real (without class loading in) in the `SingleShotTime` benchmark:
```
Benchmark                     (capacity)  (eventExecutors)  Mode  Cnt      Score   Error  Units
PoolReserveBenchmark.reserve         128                 0    ss    2     69.240          us/op
PoolReserveBenchmark.reserve         128                 1    ss    2    127.504          us/op
PoolReserveBenchmark.reserve         128                 2    ss    2    118.445          us/op
PoolReserveBenchmark.reserve         128                 4    ss    2    135.010          us/op
PoolReserveBenchmark.reserve        1024                 0    ss    2    360.428          us/op
PoolReserveBenchmark.reserve        1024                 1    ss    2    512.362          us/op
PoolReserveBenchmark.reserve        1024                 2    ss    2    513.437          us/op
PoolReserveBenchmark.reserve        1024                 4    ss    2    502.958          us/op
PoolReserveBenchmark.reserve      131072                 0    ss    2   6172.179          us/op
PoolReserveBenchmark.reserve      131072                 1    ss    2   9973.640          us/op
PoolReserveBenchmark.reserve      131072                 2    ss    2  10117.694          us/op
PoolReserveBenchmark.reserve      131072                 4    ss    2  10098.414          us/op
```
There's clearly a performance impact but is not as cathastrofic as it was before the changes

**Comment by franz1981** [comment]
This change could have bee way better in 2 ways:
1. by ensuring that no alien threads can use the new pool: this is sadly not true on https://github.com/Hyperfoil/Hyperfoil/blob/ac5c4c1bb1713e9bcf8c99f3351ba7e6269b8541/core/src/main/java/io/hyperfoil/core/impl/SimulationRunner.java#L367 which can start from a non event loop thread -> and we should evaluate why and if it can be fixed, too
2. "precise" stats tracking (for min/max) is causing a cache traffic which doesn't scale (even with `LongAdder` ) because for a partitioned counters (which has been measured with the `CountersScalability` benchmark) tracking min/max still requires to query to have the other partitions current value, causing an invalidation of the cache line where the partitioned counter belong, which both slow down reader and writer. 

The last point can be summarized, using the simplified `MESI` protocol like this:
- a counter is updated to the local cache, becoming `M` (Modified) - or `E` (Exclusive)
- a reader thread already have a version of the same counter, which is `I` (Invalid) given that has been updated by the writer
- the same reader thread query for a fresh new value, moving it to be `S` (Shared) for both itself and the writer cache
- while becoming `S` for both, if its value was different from the memory one, is updated on the memory too
- the writer thread modify it again twice, making it `E` (Exclusive) and `S`, notifying at the first transition to `E` the copy in the reader cache to be `I` (Invalid)

In short, the fact that a reader keep on asking for up to date values, slow down both, causing cache coherency traffic too.

**Review by lampajr — APPROVED** [review]
Thanks @franz1981 for all the details, I still need to get used to the micro-benchmarking and you already put some real good examples here!!!

I just left a couple of minors but it is ready to go IMO! :heart:

File: benchmarks/src/main/java/io/hyperfoil/core/harness/FastThreadLocalThreadHarnessExecutor.java
[the remark this answers is not in the document]
Comment (franz1981) [reply benchmarks/src/main/java/io/hyperfoil/core/harness/FastThreadLocalThreadHarnessExecutor.java:1]: Effectively is not required anymore - so I can drop it - good catch!

**Review by lampajr — APPROVED** [review]
Awesome!

Feel free to merge it @franz1981 :rocket:
